lib/Service/EncryptService.php

```diff
@@ -0,0 +1,399 @@
+		if (substr($data, -1 * $last) !== str_repeat(chr($last), $last)) {
+			return false;
+		}
+		return substr($data, 0, -1 * $last);
+	}
+
+
+	/**
+	 * Encrypt a credential
+	 *
+	 * @param Credential|array $credential the credential to decrypt
+	 * @return Credential|array
+	 */
+	public function decryptCredential($credential) {
+		return $this->handleCredential($credential, 'decrypt');
+	}
+
+	/**
+	 * Encrypt a credential
+	 *
+	 * @param Credential|array $credential the credential to encrypt
+	 * @return Credential|array
+	 * @throws \Exception
+	 */
+	public function encryptCredential($credential) {
+		return $this->handleCredential($credential, 'encrypt');
+	}
+
+
+
+	private function extractKeysFromCredential($credential){
+		$userKey = '';
+		$userSuppliedKey = '';
+		if ($credential instanceof Credential) {
+			$userSuppliedKey = $credential->getLabel();
+			$sk = $credential->getSharedKey();
+			$userKey = (isset($sk)) ? $sk : $credential->getUserId();
+		}
+		if (is_array($credential)) {
+			$userSuppliedKey = $credential['label'];
+			$userKey = (isset($credential['shared_key'])) ? $credential['shared_key'] : $credential['user_id'];
+		}
+		return array($userKey, $userSuppliedKey);
+	}
+
+	/**
+	 * Handles the encryption / decryption of a credential
+	 *
+	 * @param Credential|array $credential the credential to encrypt
+	 * @return Credential|array
+	 * @throws \Exception
+	 */
+	private function handleCredential($credential, $op) {
+		$service_function = ($op === 'encrypt') ? 'encrypt' : 'decrypt';
+
+		list($userKey, $userSuppliedKey) = $this->extractKeysFromCredential($credential);
+
+		$key = EncryptService::makeKey($userKey, $this->server_key, $userSuppliedKey);
+		foreach ($this->encrypted_credential_fields as $field) {
+			if ($credential instanceof Credential) {
+				$field = str_replace(' ', '', str_replace('_', ' ', ucwords($field, '_')));
+				$set = 'set' . $field;
+				$get = 'get' . $field;
+				$credential->{$set}($this->{$service_function}($credential->{$get}(), $key));
+			}
+
+			if (is_array($credential)) {
+				$credential[$field] = $this->{$service_function}($credential[$field], $key);
+			}
+		}
+		return $credential;
+	}
+
+	/**
+	 * Encrypt a file
+	 *
+	 * @param  File|array $file
+	 * @return File|array
+	 */
+
+	public function encryptFile($file) {
+		return $this->handleFile($file, 'encrypt');
+	}
+
+	/**
+	 * Decrypt a file
+	 *
+	 * @param  File|array $file
+	 * @return File|array
+	 */
+
+	public function decryptFile($file) {
+		return $this->handleFile($file, 'decrypt');
+	}
+
+	/**
+	 * Handles the encryption / decryption of a File
+	 *
+	 * @param File|array $file the credential to encrypt
+	 * @return File|array
+	 * @throws \Exception
+	 */
+	private function handleFile($file, $op) {
+		$service_function = ($op === 'encrypt') ? 'encrypt' : 'decrypt';
+		$userKey = '';
+		$userSuppliedKey = '';
+		if ($file instanceof File) {
+			$userSuppliedKey = $file->getSize();
```

Comment:
maybe use the file label or name instead of the size as usersuppliedkey and use size . mimetype on userkey